src/ripple/protocol/impl/STTx.cpp

```diff
@@ -307,13 +308,50 @@ isMemoOkay (STObject const& st, std::string& reason)
 
         for (auto const& memoElement : *memoObj)
         {
-            if ((memoElement.getFName() != sfMemoType) &&
-                (memoElement.getFName() != sfMemoData) &&
-                (memoElement.getFName() != sfMemoFormat))
+            auto const& name = memoElement.getFName();
+
+            if (name != sfMemoType &&
+                name != sfMemoData &&
+                name != sfMemoFormat)
             {
-                reason = "A memo may contain only MemoType, MemoData or MemoFormat fields.";
+                reason = "A memo may contain only MemoType, MemoData or "
+                         "MemoFormat fields.";
                 return false;
             }
+
+            // The raw data is stored as hex-octets, which we want to decode.
+            auto data = strUnHex (memoElement.getText ());
+
+            if (!data.second)
+            {
+                reason = "The MemoType, MemoData and MemoFormat fields may "
+                         "only contain hex-encoded data.";
+                return false;
+            }
+
+            std::string contents;
```

Comment:
You could build contents all at once and then check it. Something like: `std::string const contents(data.first.begin(), data.first.end()). Then in the loop don't push the chars one by one.